crd/nodeinfo/client.go

```diff
@@ -0,0 +1,161 @@
+		return current, err
+	}
+	if current == nil {
+		current, err = i.cli.Get(ctx, nodeinfo.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get existing nodeinfo crd")
+		}
+	}
+	if !reflect.DeepEqual(nodeinfo.Spec.Versions, current.Spec.Versions) {
+		nodeinfo.SetResourceVersion(current.GetResourceVersion())
+		previous := *current
+		current, err = i.cli.Update(ctx, nodeinfo, metav1.UpdateOptions{})
+		if err != nil {
+			return &previous, errors.Wrap(err, "failed to update existing nodeinfo crd")
+		}
+	}
+	return current, nil
+}
+
+// Client provides methods to interact with instances of the NodeInfo custom resource.
+type Client struct {
+	cli client.Client
+}
+
+// NewClient creates a new NodeInfo client around the passed ctrlcli.Client.
+func NewClient(cli client.Client) *Client {
+	return &Client{
+		cli: cli,
+	}
+}
+
+// Get returns the NodeInfo identified by the NamespacedName.
+func (c *Client) Get(ctx context.Context, key types.NamespacedName) (*v1alpha1.NodeInfo, error) {
+	nodeInfo := &v1alpha1.NodeInfo{}
+	err := c.cli.Get(ctx, key, nodeInfo)
+	return nodeInfo, errors.Wrapf(err, "failed to get nodeinfo %v", key)
+}
+
+// PatchSpec performs a server-side patch of the passed NodeInfoSpec to the NodeInfo specified by the NamespacedName.
+func (c *Client) PatchSpec(ctx context.Context, key types.NamespacedName, spec *v1alpha1.NodeInfoSpec, fieldManager string) (*v1alpha1.NodeInfo, error) {
+	obj := genPatchSkel(key)
+	obj.Spec = *spec
+	if err := c.cli.Patch(ctx, obj, client.Apply, client.ForceOwnership, client.FieldOwner(fieldManager)); err != nil {
+		return nil, errors.Wrap(err, "failed to patch nodeinfo")
+	}
+	return obj, nil
+}
+
+// UpdateSpec does a fetch, deepcopy, and update of the NodeInfo with the passed spec.
+// Deprecated: UpdateSpec is deprecated and usage should migrate to PatchSpec.
+func (c *Client) UpdateSpec(ctx context.Context, key types.NamespacedName, spec *v1alpha1.NodeInfoSpec) (*v1alpha1.NodeInfo, error) {
+	nodeinfo, err := c.Get(ctx, key)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get nodeinfo")
+	}
+	spec.DeepCopyInto(&nodeinfo.Spec)
+	if err := c.cli.Update(ctx, nodeinfo); err != nil {
+		return nil, errors.Wrap(err, "failed to update nodeinfo")
+	}
+	return nodeinfo, nil
+}
+
+// SetOwnerRef sets the controller of the NodeInfo to the given object atomically, using HTTP Patch.
+// Deprecated: SetOwnerRef is deprecated, use the more correctly named SetControllerRef.
+func (c *Client) SetOwnerRef(ctx context.Context, key types.NamespacedName, owner metav1.Object, fieldManager string) (*v1alpha1.NodeInfo, error) {
+	return c.SetControllerRef(ctx, key, owner, fieldManager)
+}
+
+// SetControllerRef sets the controller of the NodeInfo to the given object atomically, using HTTP Patch.
+func (c *Client) SetControllerRef(ctx context.Context, key types.NamespacedName, owner metav1.Object, fieldManager string) (*v1alpha1.NodeInfo, error) {
```

Comment:
Probably not needed, client can just create the NodeInfo CRD with ownership at time of creation, don't need these funcs which mutate the ownership over time

Reply:
removed